crates/storage/provider/src/providers/blockchain_provider.rs

```diff
@@ -0,0 +1,817 @@
+        self.database.sealed_headers_range(range)
+    }
+
+    fn sealed_headers_while(
+        &self,
+        range: impl RangeBounds<BlockNumber>,
+        predicate: impl FnMut(&SealedHeader) -> bool,
+    ) -> ProviderResult<Vec<SealedHeader>> {
+        self.database.sealed_headers_while(range, predicate)
+    }
+}
+
+impl<DB> BlockHashReader for BlockchainProvider2<DB>
+where
+    DB: Database,
+{
+    fn block_hash(&self, number: u64) -> ProviderResult<Option<B256>> {
+        self.database.block_hash(number)
+    }
+
+    fn canonical_hashes_range(
+        &self,
+        start: BlockNumber,
+        end: BlockNumber,
+    ) -> ProviderResult<Vec<B256>> {
+        self.database.canonical_hashes_range(start, end)
+    }
+}
+
+impl<DB> BlockNumReader for BlockchainProvider2<DB>
+where
+    DB: Database,
+{
+    fn chain_info(&self) -> ProviderResult<ChainInfo> {
+        Ok(self.canonical_in_memory_state.chain_info())
+    }
+
+    fn best_block_number(&self) -> ProviderResult<BlockNumber> {
+        Ok(self.canonical_in_memory_state.get_canonical_block_number())
+    }
+
+    fn last_block_number(&self) -> ProviderResult<BlockNumber> {
+        self.database.last_block_number()
+    }
+
+    fn block_number(&self, hash: B256) -> ProviderResult<Option<BlockNumber>> {
+        self.database.block_number(hash)
+    }
+}
+
+impl<DB> BlockIdReader for BlockchainProvider2<DB>
+where
+    DB: Database,
+{
+    fn pending_block_num_hash(&self) -> ProviderResult<Option<BlockNumHash>> {
+        Ok(self.canonical_in_memory_state.pending_block_num_hash())
+    }
+
+    fn safe_block_num_hash(&self) -> ProviderResult<Option<BlockNumHash>> {
+        Ok(self.canonical_in_memory_state.get_safe_num_hash())
+    }
+
+    fn finalized_block_num_hash(&self) -> ProviderResult<Option<BlockNumHash>> {
+        Ok(self.canonical_in_memory_state.get_finalized_num_hash())
+    }
+}
+
+impl<DB> BlockReader for BlockchainProvider2<DB>
+where
+    DB: Database,
+{
+    fn find_block_by_hash(&self, hash: B256, source: BlockSource) -> ProviderResult<Option<Block>> {
+        let block = match source {
+            BlockSource::Any => {
+                // check database first
+                let mut block = self.database.block_by_hash(hash)?;
+                if block.is_none() {
+                    // Note: it's fine to return the unsealed block because the caller already has
+                    // the hash
+                    block = self
+                        .canonical_in_memory_state
+                        .state_by_hash(hash)
+                        .map(|block_state| block_state.block().block().clone().unseal());
+                }
+                block
+            }
+            BlockSource::Pending => self
+                .canonical_in_memory_state
+                .state_by_hash(hash)
+                .map(|block_state| block_state.block().block().clone().unseal()),
+            BlockSource::Canonical => self.database.block_by_hash(hash)?,
```

Comment:
this is now a bit blurry, because the canonical in memory state is also canonical, so we need to handle this like any.
and pending should just check the pending block

Reply:
very good point, done here https://github.com/paradigmxyz/reth/pull/9656/commits/6a00ec8f834ae3fc858b92594a5accb4a290fcf9, now we handle `BlockSource::Canonical` and `BlockSource::Any` with the same code, I've changed it a bit to check the in memory representation before hitting the db, let me know wdyt. And for `BlockSource::Pending` we just return the pending block if present.